infrastructure/metrics/src/main/java/tech/pegasys/teku/infrastructure/metrics/MetricsHistogramWithCounters.java

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright 2022 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.infrastructure.metrics;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import org.HdrHistogram.Histogram;
+import org.HdrHistogram.SynchronizedHistogram;
+import org.hyperledger.besu.metrics.prometheus.PrometheusMetricsSystem;
+import org.hyperledger.besu.plugin.services.MetricsSystem;
+import org.hyperledger.besu.plugin.services.metrics.Counter;
+import org.hyperledger.besu.plugin.services.metrics.LabelledMetric;
+import org.hyperledger.besu.plugin.services.metrics.MetricCategory;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+
+public class MetricsHistogramWithCounters extends MetricsHistogram {
+  private final List<UInt64> sortedBoundaries;
+  private final Map<UInt64, String> boundariesToIntervalLabels;
+  private final Map<String, LabelledMetric<Counter>> intervalLabelsToMetrics;
+
+  public MetricsHistogramWithCounters(
+      final Histogram histogram,
+      final List<UInt64> sortedBoundaries,
+      final Map<UInt64, String> boundariesToIntervalLabels,
```

Comment:
This feels like it should be a `TreeMap` and then we could just use the methods in `NavigableMap` to efficiently find the key to use (e.g. `floorEntry`). Maybe I'm missing something though.